src/typing/type_inference_js.ml

```diff
@@ -3065,6 +3065,32 @@ and mk_proptype cx = Ast.Expression.(function
           (_, {Ast.Identifier.name = "oneOf"; _ });
          _
       };
+      arguments = [Expression (_, Array { Array.
+        elements = es
+      })];
+    } ->
+      let rec mk_one_of es t = match (es, t) with
+        | Some (Expression (loc, Literal {
+            Ast.Literal.value = Ast.Literal.String value
+          })) :: tl,
+          UnionT (r, ts) ->
+            let reason = mk_reason "string literal type" loc in
+            let string_literal = EnumT
+              (reason,
+               Flow_js.mk_object_with_map_proto cx reason
+                 (SMap.singleton value AnyT.t) (MixedT reason)) in
+            mk_one_of tl (UnionT (r, string_literal :: ts))
+        | [], _ -> t
+        | _  -> AnyT.at vloc in
+      let reason = mk_reason "oneOf" vloc in
+      mk_one_of es (UnionT (reason, []))
```

Comment:
Hey @samwgoldman - first of all, very cool that you're digging into this, thanks! One edit here would be to have the helper function simply build the list of types to plug into the union, rather than unwrapping and rewrapping a union at each step. Then here you could build a UnionT with the finished list. (That would also tighten the pattern match in the helper, since you wouldn't need a wildcard at the end for non-UnionT arguments.)